File: GCD/OC/GCD_01/GCD_01/ViewController.m
//
//  ViewController.m
//  GCD_01
//
//  Created by Rocky on 2018/2/24.
//  Copyright © 2018年 Rocky. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event{
//    [self asyncSerial];
    [self syncSerial];
//    获取全局并发队列
//    dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
}

//同步串行
-(void)syncSerial{
    
    dispatch_queue_t queue = dispatch_queue_create("com.download.serial", DISPATCH_QUEUE_SERIAL);
    
    dispatch_sync(queue, ^{
        NSLog(@"download xxxxxxx1111 %@",[NSThread currentThread]);
    });
    
    dispatch_sync(queue, ^{
        NSLog(@"download xxxxxxx2222 %@",[NSThread currentThread]);
    });
    
    dispatch_sync(queue, ^{
        NSLog(@"download xxxxxxx3333 %@",[NSThread currentThread]);
    });
}


//同步并发（任务串行）
- (void)syncConcurrent{
    
    dispatch_queue_t queue = dispatch_queue_create("com.xxxxx", DISPATCH_QUEUE_CONCURRENT);
    dispatch_sync(queue, ^{
        NSLog(@"download xxxxxxx %@",[NSThread currentThread]);
    });
    
    dispatch_sync(queue, ^{
        NSLog(@"download xxxxxxx2222 %@",[NSThread currentThread]);
    });
    
    dispatch_sync(queue, ^{
        NSLog(@"download xxxxxxx3333 %@",[NSThread currentThread]);
    });
}

// 异步串行
- (void)asyncSerial{
    
    dispatch_queue_t queue = dispatch_queue_create("com.download.serial", DISPATCH_QUEUE_SERIAL);
    
    dispatch_async(queue, ^{
        NSLog(@"download xxxxxxx1111 %@",[NSThread currentThread]);
    });
    
    dispatch_async(queue, ^{
        NSLog(@"download xxxxxxx2222 %@",[NSThread currentThread]);
    });
    
    dispatch_async(queue, ^{
        NSLog(@"download xxxxxxx3333 %@",[NSThread currentThread]);
    });
}

///一步函数+ 并发队列
- (void)asyncConcurrent{
    
    //DISPATCH_QUEUE_CONCURRENT 并发
    //DISPATCH_QUEUE_SERIAL 串行
    dispatch_queue_t queue = dispatch_queue_create("com.xxxxx", DISPATCH_QUEUE_CONCURRENT);
    
    dispatch_async(queue, ^{
        NSLog(@"download xxxxxxx %@",[NSThread currentThread]);
    });
    
    dispatch_async(queue, ^{
        NSLog(@"download xxxxxxx2222 %@",[NSThread currentThread]);
    });
    
    dispatch_async(queue, ^{
        NSLog(@"download xxxxxxx3333 %@",[NSThread currentThread]);
    });
}

- (void)asyncMain{
    
}

@end

File: GCD/Swift/GCD_01/GCD_01/ViewController.swift
//
//  ViewController.swift
//  GCD_01
//
//  Created by Rocky on 2018/2/24.
//  Copyright © 2018年 Rocky. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        syncSerial()
//        syncConcurrent()
//        asyncSerial()
//        asyncConcurrent()
//        获取全局队列
//        let queue = DispatchQueue.global()
    }
    
    ///同步串行
    func syncSerial()  {
        let queue = DispatchQueue(label: "com.download.serial", qos: .default)
        
        queue.sync {
            
            print("download xxxxx 11111 \(Thread.current)")
        }
        
        queue.sync {
            
            print("download xxxxx 22222 \(Thread.current)")
        }
        
        queue.sync {
            
            print("download xxxxx 33333 \(Thread.current)")
        }
    }
    
    ///同步并行
    func syncConcurrent() {
        let queue = DispatchQueue(label: "com.download.concurrent", qos: .default, attributes: .concurrent)
        
        queue.sync {
            
            print("download xxxxx 11111 \(Thread.current)")
        }
        
        queue.sync {
            
            print("download xxxxx 22222 \(Thread.current)")
        }
        
        queue.sync {
            
            print("download xxxxx 33333 \(Thread.current)")
        }
    }
    
    /// 异步串行
    func asyncSerial()  {
        
        let queue = DispatchQueue(label: "com.download.serial", qos: .default)
        
        queue.sync {
            
            print("download xxxxx 11111 \(Thread.current)")
        }
        
        queue.sync {
            
            print("download xxxxx 22222 \(Thread.current)")
        }
        
        queue.sync {
            
            print("download xxxxx 33333 \(Thread.current)")
        }
    }
    
    /// 异步并行
    func asyncConcurrent() -> Void {
        let queue = DispatchQueue(label: "com.download.concurrent", qos: .default, attributes: .concurrent)
        
        queue.async {
            
            print("download xxxxx 11111 \(Thread.current)")
        }
        
        queue.async {
            
            print("download xxxxx 22222 \(Thread.current)")
        }
        
        queue.async {
            
            print("download xxxxx 33333 \(Thread.current)")
        }
    }
}

